### ISSUE TITLE
Broken doltdb after deletion of the branch which had DB alias

### ISSUE THREAD

**Issue by dnesterenko** [dolthub/dolt#3636]
i have created 2 branches and connected to each directly via 'mysql://127.0.0.1:3306/mydb/feature-branch' so now i get database alias for each branch, see below:
mysql> show databases;
+----------------------+
| Database             |
+----------------------+
| information_schema   |
| test_db              |
| test_db/dev          |
| test_db/dev/201q8tw4 |
| test_db/dev/qpg-i_b5 |
+----------------------+
mysql> select name, hash from dolt_branches;
+--------------+----------------------------------+
| name         | hash                             |
+--------------+----------------------------------+
| dev          | 6ndmd8v8r0js4p5a9jg9jeefi9remd73 |
| dev/201q8tw4 | 10skl07tts09s3t3tghtsmegr4vnrlcg |
| dev/qpg-i_b5 | 8v7m7ft7qlv3j8nu2583khp2ueqb9nfa |
+--------------+----------------------------------+

show procedures produces some output...
mysql> SHOW PROCEDURE STATUS WHERE Db = 'test_db/dev/qpg-i_b5';
+----------------------+-----------------------------+-----------+
---------------------------------------------+--------------------
| Db                   | Name                        | Type      |
                                             | character_set_clien
											 

NEXT, i delete branch 'dev/qpg-i_b5' which does not remove appropriate db alias and it brakes whole engine from this point, after reconnect i could not retrieve procedures nor functions from existing branch:

mysql> delete from dolt_branches where name='dev/qpg-i_b5'
mysql> select name, hash from dolt_branches;
+--------------+----------------------------------+
| name         | hash                             |
+--------------+----------------------------------+
| dev          | 6ndmd8v8r0js4p5a9jg9jeefi9remd73 |
| dev/**201q8tw4** | 10skl07tts09s3t3tghtsmegr4vnrlcg |
+--------------+----------------------------------+
mysql> show databases;
+----------------------+
| Database             |
+----------------------+
| information_schema   |
| test_db              |
| test_db/dev          |
| test_db/dev/**201q8tw4** |
| test_db/dev/qpg-i_b5 |
+----------------------+

mysql> SHOW PROCEDURE STATUS WHERE Db = 'test_db/dev/**201q8tw4**';
**_Error 1105: branch not found_**
mysql> CALL DOLT_BRANCH('-d', 'dev/201q8tw4');
Error 1105: branch not found

**Comment by fulghum** [comment]
Hey @dnesterenko, thanks for reporting this issue. 

Some initial thoughts... 
* a minor general note... try to avoid deleting from or editing the `dolt_branches` table directly. Heads up that we'll eventually make that table and other system-tables read-only. It looks like you've already found the `dolt_branch` stored procedure which should give you everything you need to delete/copy/rename/create branches, so that's great. Prefer the stored procedure going forward if at all possible. 
* I started an internal discussion about why we show `<database>/<branch>` as distinct databases in the results of `show database`. It sounds like that's an artifact of how we implemented some internals, and it probably shouldn't be leaking out into `show database` as a distinct database. I'm happy to dig into this more and see if we can clean this up. 
* It is expected that a client connected to a server will become unusable when their branch is deleted out from under them. `dolt_branch('-d', ...)` actually has some safeguards to prevent deleting a branch that another active session is using, but in this case, because each client has a different db name (because the branch suffix is included as part of the db name), that logic isn't detecting that they're using the same database. 

I ran through this with two connections to two different branches and used `dolt_branch` to delete `branch2` from my client connected on `branch1`. As expected, the client on `branch2` became unusable, but the client on `branch1` was still able to execute commands correctly. I'll try to repro again a little differently and see if I can find anything else. If you could confirm that you saw just one client or both clients become unusable in your example above, that would be helpful to hear.

**Comment by dnesterenko** [comment]
hi @fulghum, thank you for looking into this one.
i think simplest way to reproduce it is as following:
1. create 2 branches under 'dolt sql-client'
2. connect to each of them via 'mysql://127.0.0.1:3306/mydb/feature-branch' that will create db alias for each branch
3. start 'dolt sql-client' and remove one of branches
4. restart 'dolt sql-client' and try any command listed in the end of initial message:
mysql> SHOW PROCEDURE STATUS;
Error 1105: branch not found

**Comment by fulghum** [comment]
Hey @dnesterenko , thanks for the extra details. I've been digging into this one today and have repro'd the behavior you see. 

To quickly summarize, two clients connect to a `dolt sql-server` using `<database>/<branch>` as the database. Dolt lists these as databases in the results of `show databases`. When one of those branches is deleted, the connection using that branch becomes unusable and has to be terminated, which is expected. What is **not expected** is that all future connections/reconnections from that point cannot execute certain queries, such as `show procedure status`, `use information_schema`, `select * from information_schema.columns`. I was still able to successfully perform queries against tables in user databases – it seems like the tables/schemas/commands that combine data from multiple databases are what's getting tripped up here. 

This seems very likely to be coming from the fact that the sql-server is still tracking that `<database>/<branch>` database who's branch was deleted. That would explain why `information_schema` queries and `show procedure status` queries stop working since they are trying to query all databases to collect the results. 

I have a few ideas for how to resolve this, but will need to keep digging in deeper to see the best way to fix things. We'll keep digging in and will keep you updated on how the fix is coming along. Thanks for reporting this one to us – we'll get this fixed asap. 

While we're digging in and getting this fixed... we don't want you and your team blocked by this issue, so I have a few _very_ short term workarounds to suggest while we get this resolved for you: 
* The easiest way to avoid this problem while we're fixing it is of course to not delete any branches. I don't know if this is feasible for your application or not, but wanted to include it as a hopefully easy temporary workaround option. 
* If you're able to temporarily stop using the <database>/<branch> form in your connection string, you won't hit this issue. You'll need to explicitly execute `call dolt_checkout('<branch>');` however as the first command in the sql session to select the right branch of course. It's not ideal, but I figured I'd mention it as a workaround just in case it's helpful.

**Comment by fulghum** [comment]
I spent some more time debugging through this today and confirmed that, as expected, this happens because the sql-server maintains metadata about the `db/branch` database that becomes invalid after the underlying branch is deleted. 

I hit a couple of obstacles with using the Go MySQL driver to test this in an automated repro, so that slowed me down a bit and will require some more work. The Go MySQL driver doesn't support connection strings where the database has a slash in it, and when I tried to switch to passing the database name as a connection attribute, I discovered that the Go MySQL driver also doesn't yet have support for connection attributes. 

So, instead, I switched over to running some manual tests while connected to a Dolt sql-server in the debugger and made some progress. After confirming the problem, I coded up a quick fix where the Dolt database provider will validate any databases that contain a slash and make sure the branch they refer to actually exists before returning it to a caller. After that change, I was no longer able to reproduce the issue. 

I'll keep working on this tomorrow, but wanted to share today's update and progress. I still need to: 1) clean up my prototype code fix, 2) figure out a way to automate a test for this, and 3) think a little more about some edge cases and make sure this is the right way to address this problem. 

I'll keep you update as I go and we'll get this one resolved asap for you!

**Comment by fulghum** [comment]
More progress today... I came up with a better way to keep the database/branch metadata in sync when a branch is deleted or renamed and also fixed the safeguard that prevents clients from deleting/renaming a branch in use by another session (unless the `force` option is used): https://github.com/dolthub/dolt/pull/3667

I've tested manually and confirmed the behavior. Tomorrow I'll focus on finding a good way to automate the tests.

**Comment by fulghum** [comment]
I found a way to automate the tests for this scenario and just pushed the tests up to my PR. I'll take another look at this tomorrow to clean up any loose ends and think about any additional testing needed, then I'll request some feedback from the team. Thanks for the patience while I worked through this one.

**Comment by fulghum** [comment]
@dnesterenko – I took another pass at cleaning up and simplifying the code and tests on Friday, then got the changes reviewed today, and merged into main. I went ahead and cut a [0.40.12 release](https://github.com/dolthub/dolt/releases/tag/v0.40.12) with this bug fix for you to pick up whenever you're ready to test the fix. 

Feel free to reopen if you hit any issues at all with the bug fix and thanks again for reporting this one so we could get it fixed!
